lib/nx_signal.ex

```diff
@@ -297,4 +467,155 @@ defmodule NxSignal do
+    freqs = f_sp * mels
+
+    min_log_hz = 1_000
+    min_log_mel = min_log_hz / f_sp
+
+    # numpy uses the f64 value by default
+    logstep = Nx.log(6.4) / 27
+
+    log_t = mels >= min_log_mel
+
+    # This is the same as freqs[log_t] = min_log_hz * Nx.exp(logstep * (mels[log_t] - min_log_mel))
+    # notice that since freqs and mels are indexed by the same conditional tensor, we don't
+    # need to slice either of them
+    mel_f = Nx.select(log_t, min_log_hz * Nx.exp(logstep * (mels - min_log_mel)), freqs)
+
+    fdiff = Nx.new_axis(mel_f[1..-1//1] - mel_f[0..-2//1], 1)
+    ramps = Nx.new_axis(mel_f, 1) - fftfreqs
+
+    lower = -ramps[0..(nmels - 1)] / fdiff[0..(nmels - 1)]
+    upper = ramps[2..(nmels + 1)//1] / fdiff[1..nmels]
+    weights = Nx.max(0, Nx.min(lower, upper))
+
+    enorm = 2.0 / (mel_f[2..(nmels + 1)] - mel_f[0..(nmels - 1)])
+
+    weights * Nx.new_axis(enorm, 1)
+  end
+
+  @doc """
+  Converts a given STFT time-frequency spectrum into a MEL-scale time-frequency spectrum.
+
+  See also: `stft/3`, `istft/3`, `mel_filters/1`
+
+  ## Options
+
+    * `:nfft` - Number of FFT bins
+    * `:fs` - Sampling frequency in Hz
+    * `:nmels` - Number of target MEL bins. Defaults to 128
+    * `:type` - Target output type. Defaults to `{:f, 32}`
+
+  ## Examples
+
+      iex> nfft = 16
+      iex> fs = 8.0e3
+      iex> {z, _, _} = NxSignal.stft(Nx.iota({10}), NxSignal.Windows.hann(n: 4), overlap_size: 2, nfft: nfft, fs: fs, window_padding: :reflect)
+      iex> Nx.axis_size(z, :frequencies)
+      16
+      iex> Nx.axis_size(z, :frames)
+      5
+      iex> NxSignal.stft_to_mel(z, nfft: nfft, fs: fs, nmels: 4)
+      #Nx.Tensor<
+        f32[frames: 5][mel: 4]
+        [
+          [0.2900530695915222, 0.17422175407409668, 0.18422472476959229, 0.09807997941970825],
+          [0.6093881130218506, 0.5647397041320801, 0.4353824257850647, 0.08635270595550537],
+          [0.7584103345870972, 0.7085014581680298, 0.5636920928955078, 0.179118812084198],
+          [0.8461772203445435, 0.7952491044998169, 0.6470762491226196, 0.2520409822463989],
+          [0.908548891544342, 0.8572604656219482, 0.7078656554222107, 0.3086767792701721]
+        ]
+      >
+  """
+  defn stft_to_mel(z, opts \\ []) do
+    magnitudes = Nx.abs(z) ** 2
+    filters = mel_filters(opts)
+
+    freq_size = div(opts[:nfft], 2)
+
+    real_freqs_mag = Nx.slice_along_axis(magnitudes, 0, freq_size, axis: :frequencies)
+    real_freqs_filters = Nx.slice_along_axis(filters, 0, freq_size, axis: :frequencies)
+
+    mel_spec =
+      Nx.dot(
+        real_freqs_mag,
+        [:frequencies],
+        real_freqs_filters,
+        [:frequencies]
+      )
+
+    mel_spec = Nx.reshape(mel_spec, Nx.shape(mel_spec), names: [:frames, :mel])
+
+    log_spec = Nx.log(Nx.clip(mel_spec, 1.0e-10, :infinity)) / Nx.log(10)
+    log_spec = Nx.max(log_spec, Nx.reduce_max(log_spec) - 8)
+    (log_spec + 4) / 4
+  end
+
+  defn linspace(min, max, opts \\ []) do
```

Comment:
Public or private?

Reply:
linspace is also used a lot in Bumblebee, it may be worth considering for Nx upstream as well

Reply:
+1 for that. It's a really common function, albeit a short one